Brian/blueprints and tower limit

cb299792458 opened the pull request:
A rough idea I had to encourage focus on terraforming. Open to lots of discussion, change, and balancing.

rockokw commented:
Could you elaborate in the description a bit more and maybe include some screenshots?

cb299792458 commented:
- Added `game.towerLimit`, preventing players from building more than a set number of towers
- Added `game.blueprints`, players can only build towers from blueprints that they have found (starting with arrow)
- Added "rewards" to the scoring modal, five choices from which the player can pick two
- rewards are castle health, gold, +2 tower limit, and two random blueprints
![image](https://github.com/user-attachments/assets/5488a4b0-6f53-4bbe-9cdc-0cbc114c9c7b)
![image](https://github.com/user-attachments/assets/ca41dda7-e139-4ced-aae0-aaa96b5160a4)

cb299792458 commented:
Preview here: [](https://project-brooklyn-git-brian-bluepri-89246e-cb299792458s-projects.vercel.app/?_vercel_share=xBEfrCHkHd7lwp4YCFIduRXIql3D34TP)

rockokw commented on frontend/src/components/ui/ScorePhaseModal.jsx:
```diff
@@ -1,23 +1,95 @@
+                while (game.blueprints.has(towerKeys[blueprintIndex] || newRewards.at(-1).label === towerKeys[blueprintIndex])) {
+                    blueprintIndex = Math.floor(Math.random() * towerKeys.length);
+                }
+                // newRewards.push({
+                //     label: `Blueprint: ${towerKeys[blueprintIndex]}`,
+                //     fn: () => game.blueprints.add(towerKeys[blueprintIndex])
+                // });
```
Instead of commenting this out, prolly better to set `BLUEPRINT_REWARDS = 0` since the surrounding code is also blueprint related

rockokw commented on frontend/src/Game.js:
```diff
@@ -329,6 +332,10 @@ export default class Game {
     }
+
+    getTowerCount = () => {
+        return this.getAllTowers().filter(t => !['castle', 'portal'].includes(t.name)).length;
```
Could just do `getAllTowers().length - 2`

rockokw commented on frontend/src/components/GamePage.jsx:
```diff
@@ -41,7 +41,7 @@ const GamePage = ({ gameMap, devMode = true }) => {
             onHide={() => setShowModal(false)}
         />
-        {showScoreModal && <ScorePhaseModal game={game} />}
+        <ScorePhaseModal game={game} show={showScoreModal} />
```
Dang, that modal ruining my plan

rockokw reviewed (approved):
Looks good but one concern regarding the commented out code:

It's generally not great to disable new functionality this way and leave dead code lying around, especially when it spans multiple files. It likely won't get maintained and can also get tricky to re-enable as you'd have to remember which files had code to uncomment.

Two options I recommend considering:

- Adding feature flags to `game` to control the new functionality and disabling them by default, e.g. `game.enableTowerLimits` and `game.enableBlueprints`, with the appropriate conditionals added where needed
- Duplicating the branch so that you have a backup copy of the code and removing the dead code from this branch

My vote would be for the first option. You could also consider adding the flags to the debug menu, which would make it easier for us to try it out if we wanted to experiment with it quickly

cb299792458 commented:
I like the idea of feature flags. Not directly related, but I'm getting a little anxious about `Game.js` becoming bloated and unreadable.

rockokw commented:
> I like the idea of feature flags. Not directly related, but I'm getting a little anxious about `Game.js` becoming bloated and unreadable.

Agreed, we could have a separate import for feature flags. Ideally what would be in `Game` is only what's needed for saving/loading the game state
